File: packages/shared/src/auction.ts
// Canonical winner election (frozen rule A). Every operator runs this over the identical relay-sealed
// order set and clearing price, so all produce the same winning arb — hence the same resultHash and an
// aggregatable signature. The Go operator (avs/internal/consensus/auction.go) mirrors this exactly and
// is held to it by the shared golden vectors in test/auction.test.ts.
//
// Rule A: the winner is the eligible order that maximises the LP surplus it leaves at the sealed
// clearing price. This always lands on the Pareto frontier of (pays more token0, wants less token1),
// so no other order can strictly dominate it — the exact condition the contract's _proveFraud checks
// (EigenAuctionTaskManager._proveFraud), making a correct election un-challengeable.
import type { ToBOrderT } from "./types";
import { toBStructHash } from "./resultHash";

const Q128 = 1n << 128n;

// LP surplus (token1 units, scaled by Q128) the arb leaves at the clearing price:
//   score = quantityIn * clearingPriceX128 - quantityOut * Q128
// i.e. the token0 paid in, valued in token1 at the clearing price, minus the token1 taken out.
export function scoreOrder(o: ToBOrderT, clearingPriceX128: bigint): bigint {
    return o.quantityIn * clearingPriceX128 - o.quantityOut * Q128;
}

// Elect the winning arb for a block. Only orders with a strictly positive surplus are eligible — a
// non-positive arb costs LPs and is never executed. Ties (equal score) break on the lower EIP-712
// struct hash so every operator picks the identical winner. Returns null when nothing is eligible;
// the block then commits with no arb (arbOrderHash = bytes32(0)).
export function electWinner(orders: ToBOrderT[], clearingPriceX128: bigint): ToBOrderT | null {
    let best: ToBOrderT | null = null;
    let bestScore = 0n;
    let bestHash = "";
    for (const o of orders) {
        const s = scoreOrder(o, clearingPriceX128);
        if (s <= 0n) continue;
        const h = toBStructHash(o);
        if (best === null || s > bestScore || (s === bestScore && h < bestHash)) {
            best = o;
            bestScore = s;
            bestHash = h;
        }
    }
    return best;
}

File: avs/internal/consensus/auction.go
package consensus

import (
	"bytes"
	"math/big"
)

var q128 = new(big.Int).Lsh(big.NewInt(1), 128)

// ScoreOrder is the LP surplus an arb leaves at the clearing price, in token1 units scaled by Q128:
// quantityIn*clearingPriceX128 - quantityOut*Q128. Mirrors packages/shared/src/auction.ts.
func ScoreOrder(o ToBOrder, clearingPriceX128 *big.Int) *big.Int {
	in := new(big.Int).Mul(o.QuantityIn, clearingPriceX128)
	return in.Sub(in, new(big.Int).Mul(o.QuantityOut, q128))
}

// ElectWinner applies the frozen rule A: among orders with strictly positive surplus, pick the one
// with the greatest surplus; ties break on the lower toBStructHash so every operator elects the
// identical winner. ok is false when nothing is eligible — the block then commits with no arb
// (arbOrderHash = bytes32(0)). Byte-for-byte the same decision as the TS reference electWinner.
func ElectWinner(orders []ToBOrder, clearingPriceX128 *big.Int) (winner ToBOrder, ok bool) {
	var bestScore *big.Int
	var bestHash [32]byte
	for _, o := range orders {
		s := ScoreOrder(o, clearingPriceX128)
		if s.Sign() <= 0 {
			continue
		}
		h := ToBStructHash(o)
		if !ok || s.Cmp(bestScore) > 0 || (s.Cmp(bestScore) == 0 && bytes.Compare(h[:], bestHash[:]) < 0) {
			winner, ok, bestScore, bestHash = o, true, s, h
		}
	}
	return winner, ok
}
